crates/mun_hir/src/module_tree.rs

```diff
@@ -0,0 +1,227 @@
+                        let child_module_id = modules.alloc(ModuleData {
+                            parent: Some(module_id),
+                            children: Default::default(),
+                            file: None,
+                        });
+                        modules[module_id]
+                            .children
+                            .insert(path_segment, child_module_id);
+                        child_module_id
+                    }
+                };
+            }
+
+            // Mark the found module with the current file id
+            let module = &mut modules[module_id];
+            if let Some(other_file) = module.file {
+                diagnostics.push(DuplicateModuleFile(module_id, vec![other_file, file_id]));
+            }
+
+            module.file = Some(file_id);
+        }
+
+        Arc::new(ModuleTree {
+            package,
+            diagnostics,
+            modules,
+            root,
+        })
+    }
+
+    /// Given a `RawVisibility` which describes the visibility of an item relative to a module into
+    /// a `Visibility` which describes the absolute visibility within the module tree.
+    pub(crate) fn resolve_visibility(
+        &self,
+        _db: &dyn DefDatabase,
+        original_module: LocalModuleId,
+        visibility: &RawVisibility,
+    ) -> Visibility {
+        match visibility {
+            RawVisibility::This => Visibility::Module(ModuleId {
+                package: self.package,
+                local_id: original_module,
+            }),
+            RawVisibility::Super => {
+                let parent_module_id = self[original_module].parent.unwrap_or(original_module);
+                Visibility::Module(ModuleId {
+                    package: self.package,
+                    local_id: parent_module_id,
+                })
+            }
+            RawVisibility::Package => Visibility::Module(ModuleId {
+                package: self.package,
+                local_id: self.root,
+            }),
+            RawVisibility::Public => Visibility::Public,
+        }
+    }
+
+    /// Returns the module that is defined by the specified `file`
+    pub fn module_for_file(&self, file: FileId) -> Option<LocalModuleId> {
+        self.modules.iter().find_map(|(idx, data)| {
+            if data.file == Some(file) {
+                Some(idx)
+            } else {
+                None
+            }
+        })
+    }
+}
+
+/// Given a path name, returns the module name
+fn normalize_module_name(name: Component) -> String {
+    name.as_str().replace("-", "_")
+}
+
+/// Given a relative path, returns a Vec with all the module names
+fn path_to_module_path(path: &RelativePath) -> Vec<String> {
+    if path.extension().is_none() {
+        path.components().map(normalize_module_name).collect()
+    } else if path
+        .file_stem()
+        .map(|stem| stem.to_lowercase())
+        .expect("the file has an extension so it must also have a file stem")
```

Comment:
What happens if I have a file `.mun`? Would this have an empty stem?

Reply:
From the docs:
```
/// The stem is:
///
/// * [None], if there is no file name;
/// * The entire file name if there is no embedded `.`;
/// * The entire file name if the file name begins with `.` and has no other `.`s within;
/// * Otherwise, the portion of the file name before the final `.`
```

So then the stem would be `.mun`, Ill add an error for invalid module names.